lib/std/start.zig

```diff
@@ -272,13 +272,26 @@ fn _start() callconv(.Naked) noreturn {
             \\ bstrins.d $sp, $zero, 3, 0
             \\ b %[posixCallMainAndExit]
             ,
+            // zig fmt: off
             .riscv32, .riscv64 =>
+            if (builtin.position_independent_code)
```

Comment:
Hmmm: https://github.com/bminor/glibc/blob/32328a5a1461ff88c0b1e04954e9c68b3fa7f56d/sysdeps/riscv/start.S#L78

Need to check what tags get emitted in practice for static binaries: https://github.com/riscv-non-isa/riscv-elf-psabi-doc/blob/master/riscv-elf.adoc#tag_riscv_x3_reg_usage-16-uleb128value

Reply:
I think they're saying that they need to setup `__global_pointer$` manually because it isn't exported for static binaries.

Reply:
I don't think that matters; if you look at the code, you can see that they end up doing `lla gp, __global_pointer$` regardless of the link mode. I would assume that it defaults to being a weak symbol, so that just turns into the moral equivalent of `mov gp, zero` in a static build.

Some more context in the arch-specific dynamic linker code: https://github.com/bminor/glibc/blob/32328a5a1461ff88c0b1e04954e9c68b3fa7f56d/sysdeps/riscv/dl-machine.h#L353-L357

Reply:
Ok, I just spent an hour reading various discussions on the riscv-elf-psabi-doc repo. I'm now thoroughly convinced that we always need to initialize `gp`. Even a simple C hello world built with `-static -fno-PIE -fno-PIC` (just for good measure) ends up with `__global_pointer$` defined and used in arbitrary C code.

Reply:
Makes sense. I looked over the docs as well and concur that it should always set the `gp`.